### ISSUE TITLE
`charmcraft build` silently omits files called out in `charmcraft.yaml` if executed from an external directory

### ISSUE THREAD

**Issue by ca-scribner** [canonical/charmcraft#514]
I'm finding some unexpected behaviour around specifying files to be copied into a charm via `charmcraft.yaml` and the execution directory when building/packing.  

# Working examples

## Specifying an exact file and executing `charmcraft` from the charm dir

Doing:

```
BASE=/tmp/charmcraft-file-copy
mkdir -p $BASE
cd $BASE

charmcraft init

mkdir to_copy
touch to_copy/foo
cat << EOF > charmcraft.yaml                                         
type: charm      
parts:
  copy_files:
    plugin: dump
    source: ./
    source-type: local
    prime:
      - ./to_copy/foo
EOF

charmcraft build -f .
```

produces `.charm` file that includes the `./to_copy/foo` as specified in the `charmcraft.yaml`.  

## Specifying an exact file and executing `charmcraft` from some other dir

This also works if you execute `charmcraft` from another directory, eg:
```
# (continue from previous example)
cd /tmp
charmcraft build -f $BASE
```

## Specifying a file pattern and executing `charmcraft` from the charm dir

Doing: 

```
BASE=/tmp/charmcraft-file-copy
mkdir -p $BASE
cd $BASE

charmcraft init

mkdir to_copy
touch to_copy/foo
cat << EOF > charmcraft.yaml                                         
type: charm      
parts:
  copy_files:
    plugin: dump
    source: ./
    source-type: local
    prime:
      - ./to_copy/*  # <--NOTE: using pattern
EOF

charmcraft build -f .
```

also works as expected

# Broken example

## Specifying a file pattern and executing `charmcraft` from some other dir

If we execute the previous example (with pattern for files) from a different directory, eg:
```
# (continue from previous example)
cd /tmp
charmcraft build -f $BASE
```

we encounter no errors, but we also **do not** have the requested files in the `.charm` file.  We see the same problem using `charmcraft pack -p ...`.  

My expectation would be that this pattern should work, but also if for some reason it cannot copy the files that it would break loudly rather than silently.

**Comment by ca-scribner** [comment]
Testing a bit more in other scenarios, I see other odd behaviour with `charmcraft pack`.  Executing from the charm dir using `charmcraft pack` (no `-p` specification) also results in wildcards not working but exact files work as expected.

**Comment by cmatsuoka** [comment]
This problem is caused by a misalignment in the "project directory" concept between Charmcraft and the parts processor. The latter assumes that source paths are relative to the directory it's executed from, and when Charmcraft is executed from an external directory the relative paths won't match the actual charm project directory. The difference in behavior between the glob and non-glob prime filter entry is caused by the filter processor accepting non-glob entries as specified while glob entries are matched against actual filesystem contents for expansion. The combination of wrong placement of relative directories and glob expansion results in the reported error.

**Comment by cmatsuoka** [comment]
This should be addressed in https://github.com/canonical/charmcraft/pull/525.

**Comment by cmatsuoka** [comment]
To prevent inconsistencies until the fixed version of charmcraft is available, change to the project dir before executing charmcraft if the project contains additional parts with sources specified as relative directories, instead of executing from an external directory.

**Comment by ca-scribner** [comment]
sounds good.  Thanks for the quick fix!

**Comment by sergiusens** [comment]
Fixed in #525
